### ISSUE TITLE
Error on model fit with stateful LSTM using Dataset

### ISSUE THREAD

**Issue by mimxrt** [tensorflow/tensorflow#31071]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04 LTS
- **TensorFlow installed from (source or binary)**: conda-forge
- **TensorFlow version (use command below)**: unknown 1.14.0
- **Python version**: Python 3.7.3
- **CUDA/cuDNN version**: NVIDIA-SMI 418.67, Driver Version: 418.67, CUDA Version: 10.1
- **GPU model and memory**: Quadro RTX 6000, 24190MiB
- **Exact command to reproduce**:

### Describe the problem
When feeding a `Dataset` to `Sequential.fit()` when a _stateful_ LSTM layer is included, a `TypeError` occurs:

    TypeError: 'DatasetV1Adapter' object is not subscriptable

This does not happen with a stateless LSTM (Using `stateful=False` does not produce the error). Below is a minimal example to reproduce this issue.

### Source code / logs
```
import numpy as np
import tensorflow as tf
#tf.enable_eager_execution()

train_X = np.arange(1, 1001).reshape((200, 5))
train_Y = np.array(list(map(
    lambda x: np.array([1, 0]) if x == 0 else np.array([0, 1]),
    np.random.randint(2, size=200))))

ds = tf.data.Dataset.from_tensor_slices((train_X, train_Y))

ws = 3
sh = None
st = 1
bs = 10
nu = 86
ne = 5

ds = ds.window(size=ws, shift=sh, stride=st, drop_remainder=True).flat_map(lambda x, y: tf.data.Dataset.zip((x.batch(ws), y.batch(ws)))).batch(bs, drop_remainder=True)

model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(input_shape=(ws, len(train_X[0])), batch_size=bs),
    tf.keras.layers.LSTM(nu, return_sequences=True, stateful=True),
    tf.keras.layers.Dense(2), # categorical
    tf.keras.layers.Activation('softmax'), # categorical
])
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['categorical_accuracy'])
print(model.summary())

history = {}
for e in range(ne):
    history[e] = model.fit(ds, epochs=1, shuffle=False)
    model.reset_states()

#history = model.fit(ds, epochs=ne, shuffle=False)
```
Output:
```
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
lstm_1 (LSTM)                (10, 3, 86)               31648     
_________________________________________________________________
dense_1 (Dense)              (10, 3, 2)                174       
_________________________________________________________________
activation_1 (Activation)    (10, 3, 2)                0         
=================================================================
Total params: 31,822
Trainable params: 31,822
Non-trainable params: 0
_________________________________________________________________
None


TypeErrorTraceback (most recent call last)
<ipython-input-2-0cef961435e5> in <module>
     30 history = {}
     31 for e in range(ne):
---> 32     history[e] = model.fit(ds, epochs=1, shuffle=False)
     33     model.reset_states()
     34 

/ws/miniconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    707         steps=steps_per_epoch,
    708         validation_split=validation_split,
--> 709         shuffle=shuffle)
    710 
    711     # Prepare validation data.

/ws/miniconda3/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py in _standardize_user_data(self, x, y, sample_weight, class_weight, batch_size, check_steps, steps_name, steps, validation_split, shuffle, extract_tensors_from_dataset)
   2707       # Check that for stateful networks, number of samples is a multiple
   2708       # of the static batch size.
-> 2709       if x[0].shape[0] % batch_size != 0:
   2710         raise ValueError('In a stateful network, '
   2711                          'you should only pass inputs with '

TypeError: 'DatasetV1Adapter' object is not subscriptable
```

**Comment by anush-o** [comment]
Issue is replicating with TF version-1.14, please find the gist of [Colab](https://colab.sandbox.google.com/drive/1N7agqrl9m2MTFK008TtJnxJA4vfBDBrF#scrollTo=-tSmJM4LQnUr). Thanks!

**Comment by jvishnuvardhan** [comment]
@mimxrt I could reproduce the issue with tf-nightly. Here is the [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/a2e382b13204b6852729ac48b1fae6d1/tf-31071_nightly.ipynb).

However, TF2.0b1 is running without any issues. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/7547bf0796f78ea7f8078b1b396aab3c/tf20_31071.ipynb). Thanks!

**Comment by mimxrt** [comment]
Thank you for testing this - good to know. However, I am unsure if I should migrate to TF2.0b1 now given it is still a beta version. To be fair, I was expecting 1.14 to be the more stable version of the two.

**Comment by qlzh727** [comment]
I think the issue is fixed recently in beee660bfcc977bb65ca0c3bec4a3ea7756ee597.
